### PR TITLE
Refactor code view specs & code view resolvers

### LINKED ISSUE #3271: Merge CodeHost.codeViewSpecs and CodeHost.codeViewSpecResolver into one CodeHost.findCodeViews() function
The two ways currently add complexity in handling both. This complexity will add up when we support more kinds of code view specs (e.g. for diffs).
In addition, they are less powerful, because `codeViewResolver` only allows a single selector.
On the other hand there is not much gained from letting the caller run `resolver.resolveCodeViewSpec(querySelectorAll(resolver.selector))` instead of just letting the resolver (a single `findCodeViews()` function) run `querySelectorAll()` and return `ResolvedCodeView[]` found.

We could also get rid of the different `ResolvedCodeView` and `CodeView` types, there would only be the resolved version.

It currently makes it impossible to express a code view that is found by a selector targeting one of its children, which is what the single file blob view on GitHub and Phabricator needs to do.

#### Interface 
```ts
findCodeViews: (container: HTMLElement) => ResolvedCodeView[]
```

**Alternative that was considered**: Having a `CodeView.find()` method, but it is useful to have one function that can search for a common selector and then determine the kind of code view for each result (e.g. GitHub `.file` selector)

### LINKED ISSUE #3480: Code intel sometimes doesn't work on GitHub PRs
- Visit https://github.com/sourcegraph/sourcegraph/pull/3221
- Hover on some tokens
- Notice this console error:

image (15)

Original Slack discussion: https://sourcegraph.slack.com/archives/C07KZF47K/p1555544943062900

### PR DESCRIPTION
**Pull request by lguychard**
Review by commit.

Fixes #3271, #3491  

After making the changes in https://github.com/sourcegraph/sourcegraph/pull/3488, code intelligence was still broken on this PR: https://github.com/sourcegraph/sourcegraph/pull/3471

This was due to https://github.com/sourcegraph/sourcegraph/issues/3491. The best way to fix it cleanly was to make `fileLineContainerCodeView` a `ViewResolver`. While I was at it I merged `CodeHost.codeViewSpecs` and `CodeHost.codeViewResolver` into `CodeHost.codeViewResolvers: ViewResolver<CodeView>`.

### Code Hosts

- [x] GitHub
- [x] Gitlab

### Browsers

- [x] Chrome
- [x] Firefox

### CODE REVIEW

File: client/browser/src/libs/github/code_intelligence.ts [hunk 1]
```diff
@@ -136,43 +136,53 @@ const commentSnippetCodeViewResolver = toCodeViewResolver('.js-comment-body', {
     toolbarButtonProps,
 })
 
+const getSingleFileToolbarMount: NonNullable<CodeView['getToolbarMount']> = (
+    repositoryContent: HTMLElement
```
Comment (felixfbecker) [inline client/browser/src/libs/github/code_intelligence.ts]: Shouldn't this be `codeViewElement`? `repositoryContent` sounds like the entire page to me

File: client/browser/src/libs/github/code_intelligence.ts [hunk 2]
```diff
@@ -24,9 +24,10 @@ import { setElementTooltip } from './tooltip'
 import { getFileContainers, parseURL } from './util'
 
 /**
- * Creates the mount element for the CodeViewToolbar.
+ * Creates the mount element for the CodeViewToolbar on code views containing
+ * a `.file-actions` element.
```
Comment (felixfbecker) [inline client/browser/src/libs/github/code_intelligence.ts]: Maybe enumerate some examples here of such code views

File: client/browser/src/libs/github/util.tsx [hunk 3]
```diff
@@ -37,39 +37,29 @@ function getPathNamesFromElement(element: HTMLElement): { headFilePath: string;
...
-    const toolbar = document.querySelector<HTMLElement>(toolbarSelector)
-    if (!toolbar) {
-        throw new Error(`Could not find out if split diff is enabled, expected toolbar ${toolbarSelector} to exist`)
+    const codeView = element.classList.contains('file') ? element : element.closest('.file')
+    if (!codeView) {
+        throw new Error('Could not resolve code view element')
+    }
+    if (codeView.classList.contains('js-comment-container')) {
+        // Commented snippet in PR discussion
+        return true
```
[the remark this answers is not in the document]
Comment (lguychard) [reply client/browser/src/libs/github/util.tsx]: This is outdated and was fixed in https://github.com/sourcegraph/sourcegraph/pull/3494/commits/5df4c8fed836038c4dd99b23182d39b50f2c11da

File: client/browser/src/libs/phabricator/code_intelligence.ts [hunk 4]
```diff
@@ -126,48 +125,43 @@ export const diffCodeView = {
...
-        dom: diffusionDOMFns,
-        resolveFileInfo: resolveDiffusionFileInfo,
-        getToolbarMount: () => {
-            const actions = document.querySelector<HTMLElement>('.phui-two-column-content .phui-header-action-links')
-            if (!actions) {
-                throw new Error('unable to find file actions')
-            }
+// TODO this code view does not include the toolbar,
+// which makes it not possible to test getToolbarMount()
+// Fix after https://github.com/sourcegraph/sourcegraph/issues/3271
```
[the remark this answers is not in the document]
Comment (lguychard) [reply client/browser/src/libs/phabricator/code_intelligence.ts]: This PR closes that issue, but does not refactor this specific phabricator code view to include the toolbar.

File: client/browser/src/libs/github/util.tsx [hunk 5]
```diff
@@ -37,39 +37,29 @@ function getPathNamesFromElement(element: HTMLElement): { headFilePath: string;
...
-}
-
 /**
  * Returns if the current view shows diffs with split (vs. unified) view.
+ *
+ * @param element, either an element contained in a code view or the code view itself
  */
-export function isDomSplitDiff(): boolean {
-    const { isDelta, isPullRequest } = parseURL()
+export function isDomSplitDiff(element: HTMLElement): boolean {
```
[the remark this answers is not in the document]
Comment (lguychard) [reply client/browser/src/libs/github/util.tsx]: Done in https://github.com/sourcegraph/sourcegraph/pull/3571/commits/1e388a6f1c823de9bcf8dc115204dd58f6c97c4c

File: client/browser/src/libs/github/code_intelligence.test.ts [hunk 6]
```diff
@@ -13,10 +13,12 @@ describe('github/code_intelligence', () => {
                             for (const view of ['split', 'unified']) {
```
[the remark this answers is not in the document]
Comment (lguychard) [reply client/browser/src/libs/github/code_intelligence.test.ts]: Done in https://github.com/sourcegraph/sourcegraph/pull/3571/commits/a01d52da4feabc3c849ec1755d0cc61fc0b70c8c

File: client/browser/src/libs/code_intelligence/code_views.ts [hunk 7]
```diff
@@ -10,12 +10,10 @@ import { ensureRevisionsAreCloned } from './util/file_info'
...
 /**
- * Defines a type of code view a given code host can have. It tells us how to
- * look for the code view and how to do certain things when we find it.
+ * Describes a set of operations for manipulating a code view found on a page,
+ * and some CSS classes to be applied to injected UI elements.
  */
-export interface CodeViewSpec {
-    /** A selector used by `document.querySelectorAll` to find the code view. */
-    selector: string
+export interface CodeViewSpec extends Pick<CodeView, Exclude<keyof CodeView, 'element'>> {
```
[the remark this answers is not in the document]
Comment (lguychard) [reply client/browser/src/libs/code_intelligence/code_views.ts]: Done in https://github.com/sourcegraph/sourcegraph/pull/3571/commits/932e6f4d27cdbb17e238139a431a145c35147eaf

File: client/browser/src/libs/github/code_intelligence.ts (hunk 1 above)
Comment (lguychard) [reply client/browser/src/libs/github/code_intelligence.ts]: Done in https://github.com/sourcegraph/sourcegraph/pull/3571/commits/f4289bfedfb840be88306955d90125a705377707

File: client/browser/src/libs/github/code_intelligence.ts (hunk 2 above)
Comment (lguychard) [reply client/browser/src/libs/github/code_intelligence.ts]: Done in https://github.com/sourcegraph/sourcegraph/pull/3571/commits/78651284805d588b307a4a75ec054551da9e5836

File: client/browser/src/libs/github/dom_functions.ts [hunk 8]
```diff
@@ -15,6 +11,10 @@ const getDiffCodePart = (codeElement: HTMLElement): DiffPart => {
     if (td.classList.contains('blob-code-deletion')) {
         return 'base'
     }
+    if (isDomSplitDiff()) {
```
[the remark this answers is not in the document]
Comment (lguychard) [reply client/browser/src/libs/github/dom_functions.ts]: Improved comment in https://github.com/sourcegraph/sourcegraph/pull/3571/commits/3b58b899015a27036d18d47238cd9abf6d456e05

**Pull request merged**